proof/refine/ARM/Tcb_R.thy

```diff
@@ -24,47 +24,262 @@ lemma activate_idle_thread_corres:
+  by (clarsimp simp: invs'_def valid_state'_def valid_idle'_def pred_tcb_at'_def obj_at'_def
+                     idle_tcb'_def)
+
+(* FIXME RT: move *)
+lemma maybeM_when:
+  "maybeM f x = when (x \<noteq> None) (f (the x))"
+  unfolding maybeM_def
+  by (clarsimp split: option.splits)
+
+lemma corres_assert_opt_assume_l'':
+  "corres_underlying sr nf nf' r P Q (f (the X)) g
+  \<Longrightarrow> corres_underlying sr nf nf' r (P and K (X \<noteq> None)) Q (assert_opt X >>= f) g"
+  by (force simp: corres_underlying_def assert_opt_def return_def bind_def fail_def)
+
+lemma scConsumed_update_corres:
+ "x = y \<Longrightarrow>
+  corres dc (sc_at scp) (ko_at' sc scp)
+    (set_sc_obj_ref sc_consumed_update scp x) (setSchedContext scp (scConsumed_update (\<lambda>_. y) sc))"
+  unfolding update_sched_context_def
+  apply (rule corres_guard_imp)
+    apply (rule corres_symb_exec_l'[where Q'="\<lambda>r s. ko_at r scp s \<and>  (\<exists>n. is_sc_obj n r)",
+                                    where P="\<lambda>s. obj_at \<top> scp s"])
+      apply (rule corres_guard_imp[where P'=R and Q'=R for R])
+        apply (rule_tac F="(\<exists>n. is_sc_obj n obj)" in corres_gen_asm)
+        apply (case_tac obj; simp add: is_sc_obj_def)
+        apply (rule setSchedContext_no_stack_update_corres[where f'="scConsumed_update (\<lambda>_. y)"])
+           apply (clarsimp simp: sc_relation_def objBits_def objBitsKO_def obj_at_def)+
+     apply (fastforce simp: exs_valid_def get_object_def in_monad obj_at_def)
+    apply (wpsimp wp: get_object_wp)
+   apply (fastforce simp: obj_at_def)
+  apply simp
+  done
+
+lemma schedContextUpdateConsumed_corres:
+ "corres (=) (sc_at scp) (sc_at' scp)
+            (sched_context_update_consumed scp)
+            (schedContextUpdateConsumed scp)"
+  unfolding sched_context_update_consumed_def schedContextUpdateConsumed_def
+  apply (simp add: maxTicksToUs_def ticksToUs_def)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split [OF _ get_sc_corres])
+      apply (rule corres_if2, clarsimp simp: maxTicksToUs_def sc_relation_def)
+       apply (rule corres_split[OF corres_return_eq_same[OF refl]])
+         apply (rule scConsumed_update_corres, clarsimp simp: sc_relation_def)
+        apply wpsimp+
+      apply (rule corres_split[OF corres_return_eq_same scConsumed_update_corres],
+             clarsimp simp: sc_relation_def)
+        apply wpsimp+
+  done
+
+crunches sched_context_update_consumed
+  for in_user_frame[wp]: "in_user_frame x"
+
+lemma fhhfhf[wp]:
+  "schedContextUpdateConsumed scp \<lbrace>\<lambda>s. tcb_at' (ksCurThread s) s\<rbrace>"
+  unfolding schedContextUpdateConsumed_def
+  by (wpsimp | wps)+
+
+lemma fffhhfhf[wp]:
+  "schedContextUpdateConsumed scp \<lbrace>valid_ipc_buffer_ptr' x\<rbrace>"
+  unfolding schedContextUpdateConsumed_def valid_ipc_buffer_ptr'_def
+  by wpsimp
+
+(* this trivial lemma is used to resolve schematic variables in the case that we want
```

Comment:
I would like to remove this if possible, but not sure of the best way to remove it.